File: taurus-protect-sdk-go/pkg/protect/model/exchange.go
package model

import "time"

// Exchange represents an exchange account in the Taurus-PROTECT system.
type Exchange struct {
	// ID is the unique identifier for the exchange account.
	ID string `json:"id"`
	// Exchange is the name of the exchange (e.g., "Binance", "Coinbase").
	Exchange string `json:"exchange"`
	// Account is the account identifier on the exchange.
	Account string `json:"account"`
	// Currency is the currency identifier for this exchange account.
	Currency string `json:"currency"`
	// Type is the type of exchange account.
	Type string `json:"type"`
	// TotalBalance is the total balance in the smallest currency unit.
	TotalBalance string `json:"total_balance"`
	// Status is the status of the exchange account.
	Status string `json:"status"`
	// Container is the container identifier.
	Container string `json:"container"`
	// Label is a user-defined label for the exchange account.
	Label string `json:"label"`
	// DisplayLabel is the display label for the exchange account.
	DisplayLabel string `json:"display_label"`
	// HasWLA indicates if the exchange account has a whitelisted address.
	HasWLA bool `json:"has_wla"`
	// BaseCurrencyValuation is the valuation in the base currency (e.g., CHF, EUR, USD).
	BaseCurrencyValuation string `json:"base_currency_valuation,omitempty"`
	// CurrencyInfo contains detailed information about the currency.
	CurrencyInfo *Currency `json:"currency_info,omitempty"`
	// CreationDate is when the exchange account was created.
	CreationDate time.Time `json:"creation_date"`
	// UpdateDate is when the exchange account was last updated.
	UpdateDate time.Time `json:"update_date"`
}

// ExchangeCounterparty represents an exchange counterparty with its total valuation.
type ExchangeCounterparty struct {
	// Name is the name of the exchange.
	Name string `json:"name"`
	// BaseCurrencyValuation is the total valuation in the base currency.
	BaseCurrencyValuation string `json:"base_currency_valuation"`
}

// ListExchangesOptions contains options for listing exchange accounts.
type ListExchangesOptions struct {
	// CurrencyID filters by currency ID.
	CurrencyID string
	// IncludeBaseCurrencyValuation includes the base currency valuation in results.
	IncludeBaseCurrencyValuation bool
	// ExchangeLabel filters by exchange label.
	ExchangeLabel string
	// SortOrder is the sort order (ASC or DESC). Default is DESC.
	SortOrder string
	// Status filters by exchange account status.
	Status string
	// OnlyPositiveBalance excludes accounts with zero balance when true.
	OnlyPositiveBalance bool
	// CurrentPage is the base64-encoded cursor for the current page.
	CurrentPage string
	// PageRequest indicates which page to request (FIRST, PREVIOUS, NEXT, LAST).
	PageRequest string
	// PageSize is the number of items per page.
	PageSize int64
}

// ListExchangesResult contains the result of listing exchange accounts.
type ListExchangesResult struct {
	// Exchanges is the list of exchange accounts.
	Exchanges []*Exchange `json:"exchanges"`
	// CurrentPage is the base64-encoded cursor for the current page.
	CurrentPage string `json:"current_page,omitempty"`
	// HasPrevious indicates if there is a previous page.
	HasPrevious bool `json:"has_previous"`
	// HasNext indicates if there is a next page.
	HasNext bool `json:"has_next"`
}

// ListExchangeCounterpartiesResult contains the result of listing exchange counterparties.
type ListExchangeCounterpartiesResult struct {
	// ExchangesTotalValuation is the total valuation of all exchanges in the base currency.
	ExchangesTotalValuation string `json:"exchanges_total_valuation"`
	// Exchanges is the list of exchange counterparties.
	Exchanges []*ExchangeCounterparty `json:"exchanges"`
}

// ExportExchangesOptions contains options for exporting exchange accounts.
type ExportExchangesOptions struct {
	// Format specifies the export format ("csv" or "json").
	Format string
}

// ExportExchangesResult contains the result of exporting exchange accounts.
type ExportExchangesResult struct {
	// Result contains the exported data (CSV or JSON string).
	Result string `json:"result"`
	// TotalItems is the total number of items exported.
	TotalItems int64 `json:"total_items"`
}

File: taurus-protect-sdk-typescript/src/models/exchange.ts
/**
 * Exchange models for Taurus-PROTECT SDK.
 *
 * Represents exchange account information in the Taurus-PROTECT system.
 */

import type { Currency } from './currency';
import type { CursorNavigationOptions, CursorPage } from './pagination';

/**
 * Represents an exchange account in the Taurus-PROTECT system.
 *
 * An exchange represents a connection to a third-party exchange (e.g., Binance,
 * Coinbase) that can be used for trading and transfers.
 */
export interface Exchange {
  /** Unique identifier */
  readonly id?: string;
  /** Exchange name (e.g., "binance", "coinbase") */
  readonly exchange?: string;
  /** Account identifier on the exchange */
  readonly account?: string;
  /** Currency code */
  readonly currency?: string;
  /** Account type */
  readonly type?: string;
  /** Total balance in the smallest currency unit */
  readonly totalBalance?: string;
  /** Account status */
  readonly status?: string;
  /** Container */
  readonly container?: string;
  /** Label */
  readonly label?: string;
  /** Display label */
  readonly displayLabel?: string;
  /** Valuation in the base currency (CHF, EUR, USD, etc.) */
  readonly baseCurrencyValuation?: string;
  /** Whether the exchange has a whitelisted address */
  readonly hasWLA?: boolean;
  /** Currency information */
  readonly currencyInfo?: Currency;
  /** Creation date */
  readonly creationDate?: Date;
  /** Last update date */
  readonly updateDate?: Date;
}

/**
 * Represents an exchange counterparty summary.
 *
 * A counterparty represents a grouped view of exchange holdings by exchange name,
 * with the total valuation across all currencies.
 */
export interface ExchangeCounterparty {
  /** Exchange name */
  readonly name?: string;
  /** Total valuation in the base currency (CHF, EUR, USD, etc.) */
  readonly baseCurrencyValuation?: string;
}

/**
 * Represents the withdrawal fee for an exchange transfer.
 *
 * Contains the fee amount that will be charged for withdrawing
 * assets from an exchange to an external address.
 */
export interface ExchangeWithdrawalFee {
  /** Withdrawal fee amount in the smallest currency unit */
  readonly fee?: string;
}

/**
 * Options for listing exchange accounts. A cursor list: `pageSize` 1-100 (default 20) and
 * the `cursor` of a previous page.
 */
export interface ListExchangesOptions extends CursorNavigationOptions {
  /** Filter on currency ID */
  currencyId?: string;
  /** Include base currency valuation in response */
  includeBaseCurrencyValuation?: boolean;
  /** Filter by exchange label */
  exchangeLabel?: string;
  /** Sort order: "ASC" or "DESC" (default: "DESC") */
  sortOrder?: 'ASC' | 'DESC';
  /** Filter by status (leave empty for all statuses) */
  status?: string;
  /** Exclude exchange accounts with zero balance */
  onlyPositiveBalance?: boolean;
}

/**
 * A page of exchange accounts.
 */
export interface ListExchangesResult {
  /** The exchange accounts of this page */
  items: Exchange[];
  /** Cursor pagination: continue with `cursor: pagination.nextCursor` while `hasMore` */
  pagination: CursorPage;
}

/**
 * Options for getting withdrawal fee.
 */
export interface GetWithdrawalFeeOptions {
  /** Destination address ID (optional) */
  toAddressId?: string;
  /** Amount to withdraw in the smallest currency unit (optional) */
  amount?: string;
}
